src/wallet/api/wallet.cpp

```diff
@@ -2016,7 +2036,8 @@ void WalletImpl::doRefresh()
         }
     } catch (const std::exception &e) {
         setStatusError(e.what());
-    }
+    }while(!rescan && (rescan=m_refreshShouldRescan.exchange(false))); // repeat if not rescanned and rescan was requested
```

Comment:
Infinite loop

Reply:
The loop will stop when no more rescans are requested from API thread via m_refreshShouldRescan atomic

Reply:
It won't stop

Reply:
It seemed OK when I looked. What is the case which makes it infinite loop ?

Reply:
rescan=true: will exit since we have false in while condition
rescan=false: will write false to rescan and will exit on next iteration of the loop since its condition will evaluate to false. This corresponds to additional rescans requested while in refresh

Both cases will gracefully exit the loop

Reply:
Just call `doRefresh()` with `m_refreshShouldRescan == false`

Reply:
in this case rescan=false and if m_refreshShouldRescan is still false the loop will exit. It will continue only if m_refreshShouldRescan is assigned true while we're in the loop. In this case it will exit one iteration later.

It could loop infinitely only if someone infinitely requests `rescanBlockchainAsync()` from API thread. But nobody would normally do it.

Did you already applied some unit test  which resulted in such infinite loop? Could I see the code of this test please?

Reply:
Did you even test the code?

Reply:
Just explain your test case precisely, so then either you will realize it's OK, or we will realize it's not :) AFAICT the code's OK.

Reply:
Seems i've missed something, will take a look again